src/librustdoc/html/render.rs

```diff
@@ -1363,21 +1404,42 @@ impl Context {
         }
 
         if item.is_mod() {
-            // modules are special because they add a namespace. We also need to
+            // Modules are special because they add a namespace. We also need to
             // recurse into the items of the module as well.
             let name = item.name.as_ref().unwrap().to_string();
             let mut item = Some(item);
+
             self.recurse(name, |this| {
                 let item = item.take().unwrap();
 
                 let mut buf = Vec::new();
-                render(&mut buf, this, &item, false).unwrap();
-                // buf will be empty if the module is stripped and there is no redirect for it
+                this.render_item(&mut buf, &item, false).unwrap();
+
+                // buf will be empty if the module is stripped and there is no redirect for it.
                 if !buf.is_empty() {
-                    let joint_dst = this.dst.join("index.html");
                     try_err!(fs::create_dir_all(&this.dst), &this.dst);
+                    // The surrounding call to `recurse` adds the modules name to this.dst, so we
+                    // only need to add 'index.html' to get the correct URL for the module.
+                    let joint_dst = this.dst.join("index.html");
                     let mut dst = try_err!(File::create(&joint_dst), &joint_dst);
                     try_err!(dst.write_all(&buf), &joint_dst);
+
+                    // Write a redirect from foo/bar/baz.t.html to foo/bar/baz/index.html
+                    // since both are sensible ways to try and find a module's
+                    // page. If a page exists already, do not overwrite it. We
+                    // don't want to replace a real rustdoc page with a redirect.
+                    // In particular, the redirect for a primitive module would
+                    // overwrite the doc page for the primitive itself.
```

Comment:
> If these redirects don't necessarily exist, is there much point in them existing at all?

I would have thought that 2% of links breaking is better than 100% of links breaking.

> The following documents fine currently but doesn't with this PR.

This seems to be a totally hypothetical question - is there any reason for anyone to actually write this?

> I think it might make sense to put primitives in their own "namespace" because they're special

I thought about this, but it kind of ruins the whole idea - you can't just look up a type, you need to know if that type is a primitive or not - how do you know that? There is no way to tell, you need to look at a magic list that will go out of date as soon as we add `i128` or whatever.

Furthermore, they're not special, they're only special to Rustdoc, which has an ugly hack to support documenting them, that should be an implementation detail, not exposed in the URLs.